### PR TITLE
Pubdev 2535 Pivot v2

### PR DESCRIPTION
**Pull request by mklechan**
This is a much improved version of Pivot.  First does a sort by index.  Then pivots locally within Chunks, collapsing rows sequentially as the index is ordered.

The cleanup mrtask then merges identical indexes and crosses Chunk boundaries.  Checks to see if the current Chunk has a run that continues from the previous Chunk.

This method scales with column cardinality much better than the previous method.

Limitation is that all rows of a single index value needs to fit on one node 
and max rows for a single index value and column label is Chunk size * Chunk size

Todo: More thorough checking on the pyunit.  Clean up some comments in AstPivot.java

### CODE REVIEW

**Comment by micahstubbs** [comment]
assigning to @navdeep-G for review, per the conversation on slack

**Comment by mklechan** [comment]
@h2o-ops please test!

**Comment by mklechan** [comment]
@h2o-ops please test!

**Comment by mklechan** [comment]
Could someone take a look at this PR?

**Comment by mmalohlava** [comment]
Sorry Mark, it is in the queue

**Comment by navdeep-G** [comment]
I will look today.

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  public String str() {
+    return "pivot";
+  }
+
+  @Override
+  public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
+    // Distributed parallelized mrtask pivot
+    // Limitations: a single index value cant have more than chunk size * chunk size number of rows
+    //  or if all rows of a single index value cant fit on a single node (due to the sort call)
+    Frame fr = stk.track(asts[1].exec(env)).getFrame();
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java]: Maybe you can inherit from AstBuiltin, it will extract the argument for you.

Comment (navdeep-G) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java]: Yes, thats a good idea.

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    }
+    @Override
+    public void map(Chunk[] cs, NewChunk[] nc) {
+      for (int i = 0; i < cs[0]._len; i++) {
+        // consolidate locally. we know its sorted by index
+        // i will be jumping forward
+        long currentIdx = cs[_indexColIdx].at8(i); // the row label
+        int count = 0; // the secondary chunk row counter
+        double[] newRow = new double[nc.length-1];   // the pivot accumulator
+        for (int j = 0; j < nc.length-1; j++) newRow[j] = Double.NaN;
```
Comment (navdeep-G) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java]: Why not use `Arrays.fill(newRow,Double.NaN);` instead of a loop?

File: h2o-py/h2o/frame.py [hunk 3]
```diff
@@ -2343,6 +2343,31 @@ def isax(self, num_words, max_cardinality, optimize_card=False, **kwargs):
         if max_cardinality <= 0: raise H2OValueError("max_cardinality must be greater than 0")
         return H2OFrame._expr(expr=ExprNode("isax", self, num_words, max_cardinality, optimize_card))
 
+    def pivot(self, index, column, value):
+        """
+        Pivot the frame designated by the three columns: index, column, and value
+        Index is a column that will be the row label
+        Column is a column that contains categorical levels which will each be turned into a column
+        Value is a column associated with an index and column
```
Comment (navdeep-G) [inline h2o-py/h2o/frame.py]: Should lower case all parameters here. Also, no need to explain what each parameter is at this point. Save that for `:param`.

File: h2o-py/h2o/frame.py [hunk 4]
```diff
@@ -2343,6 +2343,31 @@ def isax(self, num_words, max_cardinality, optimize_card=False, **kwargs):
         if max_cardinality <= 0: raise H2OValueError("max_cardinality must be greater than 0")
         return H2OFrame._expr(expr=ExprNode("isax", self, num_words, max_cardinality, optimize_card))
 
+    def pivot(self, index, column, value):
+        """
+        Pivot the frame designated by the three columns: index, column, and value
+        Index is a column that will be the row label
```
Comment (navdeep-G) [inline h2o-py/h2o/frame.py]: What are the proper values for index? Time?

File: h2o-py/h2o/frame.py [hunk 5]
```diff
@@ -2343,6 +2343,31 @@ def isax(self, num_words, max_cardinality, optimize_card=False, **kwargs):
...
+        Column is a column that contains categorical levels which will each be turned into a column
+        Value is a column associated with an index and column
+
+        :param index:
+        :param column:
+        :param value:
+        :return:
+        """
+        assert_is_type(index, str)
```
Comment (navdeep-G) [inline h2o-py/h2o/frame.py]: The index can be anything as long as its passed in as a string?

File: h2o-py/tests/testdir_misc/pyunit_pivot_medium.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    df = h2o.create_frame(rows=1000000,
+                          cols=3,
+                          factors=10,
+                          categorical_fraction=1.0/3,
+                          time_fraction=1.0/3,
+                          real_fraction=1.0/3,
+                          real_range=100,
+                          missing_fraction=0.0,
+                          seed=123)
+    dcol = df.moment(year=df["C1"].year(),month=df["C1"].month(),day=df["C1"].day())
```
Comment (navdeep-G) [inline h2o-py/tests/testdir_misc/pyunit_pivot_medium.py]: Good you are making a date here. But how would this do if the column is already a date? Assuming the user would have to call `moments` to make a time column H2O can handle?

File: h2o-py/tests/testdir_misc/pyunit_pivot_medium.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        df1 = h2o.H2OFrame({"index":range(1,s+1),"column":["a"]*s,"value":[1]*s})
+        df2 = h2o.H2OFrame({"index":range(1,s+1),"column":["b"]*s,"value":[2]*s})
+        df3 = h2o.H2OFrame({"index":range(1,s+1),"column":["c"]*s,"value":[3]*s})
+        dfall = df1.rbind(df2)
+        dfall = dfall.rbind(df3)
+        res = dfall.pivot(index="index", column="column", value="value")
+        assert res["a"].sum()[0,0] == 1*s
+        assert res["b"].sum()[0,0] == 2*s
+        assert res["c"].sum()[0,0] == 3*s
+        h2o.remove_all()
```
Comment (navdeep-G) [inline h2o-py/tests/testdir_misc/pyunit_pivot_medium.py]: No need to `remove_all()` here.

File: h2o-py/tests/testdir_misc/pyunit_pivot_medium.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        print(str(s))
+        df1 = h2o.H2OFrame({"index":range(1,s+1),"column":["a"]*s,"value":[1]*s})
+        df2 = h2o.H2OFrame({"index":range(1,s+1),"column":["b"]*s,"value":[2]*s})
+        df3 = h2o.H2OFrame({"index":range(1,s+1),"column":["c"]*s,"value":[3]*s})
+        dfall = df1.rbind(df2)
+        dfall = dfall.rbind(df3)
+        res = dfall.pivot(index="index", column="column", value="value")
+        assert res["a"].sum()[0,0] == 1*s
+        assert res["b"].sum()[0,0] == 2*s
+        assert res["c"].sum()[0,0] == 3*s
```
Comment (navdeep-G) [inline h2o-py/tests/testdir_misc/pyunit_pivot_medium.py]: Good to add an exception message here if the `assert` fails.

File: h2o-py/tests/testdir_misc/pyunit_pivot_medium.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+                          real_range=100,
+                          missing_fraction=0.0,
+                          seed=123)
+    dcol = df.moment(year=df["C1"].year(),month=df["C1"].month(),day=df["C1"].day())
+    df2 = df.cbind(dcol)
+    df2.show()
+    df3 = df2.pivot(index="time", column="C2", value="C3")
+    df3.show()
+    assert len(df3) == 18250
+    assert len(df3.columns) == 11
```
Comment (navdeep-G) [inline h2o-py/tests/testdir_misc/pyunit_pivot_medium.py]: Good to add an assert message here if the check fails.

File: h2o-r/tests/testdir_munging/runit_pivot.R [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+                         integer_fraction=1.0/3,
+                         categorical_fraction=1.0/3,
+                         missing_fraction=0.0,
+                         seed=123)
+    df$C3 = h2o.abs(df$C3)
+    head(df)
+    df2 = h2o.pivot(df,index="C3",column="C2",value="C1")
+    head(df2)
+    expect_equal(nrow(df2),101)
+    expect_equal(ncol(df2),11)
```
Comment (navdeep-G) [inline h2o-r/tests/testdir_munging/runit_pivot.R]: This test seems incomplete compared to the Python unit test.

File: h2o-r/tests/testdir_munging/runit_pivot.R [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+                         cols=3,
+                         factors=10,
+                         integer_fraction=1.0/3,
+                         categorical_fraction=1.0/3,
+                         missing_fraction=0.0,
+                         seed=123)
+    df$C3 = h2o.abs(df$C3)
+    head(df)
+    df2 = h2o.pivot(df,index="C3",column="C2",value="C1")
+    head(df2)
```
Comment (navdeep-G) [inline h2o-r/tests/testdir_munging/runit_pivot.R]: Why print `head()` here?

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+          // for now we are aggregating by "first"
+          for (int j = 0; j < nc.length - 1; j++) {
+            if (Double.isNaN(newRow[j]) && !Double.isNaN(cs[j + 1].atd(i + count))) {
+              newRow[j] = cs[j + 1].atd(i + count);
+            }
+          }
+          // need to look if we need to go to next chunk
+          if (i + count == cs[0]._len - 1) {
+            Chunk nextChunk = cs[0].nextChunk();
+            if (nextChunk != null && currentIdx == nextChunk.at8(0)) {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java]: What will happen if one "index" spans more than 2 chunks? (eg. the index column is constant)?

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java (hunk 1 above)
Comment (mklechan) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java]: :+1:

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    Frame result = new Frame(secondPass.vec(0).makeCopy(null,fr.vec(indexIdx).get_type()));
+    result._key = Key.<Frame>make();
+    result.setNames(new String[]{index});
+    secondPass.remove(0);
+    result.add(secondPass);
+    initialPass.delete();
+    return new ValFrame(result);
+  }
+
+  protected class pivotCleanup extends MRTask<AstPivot.pivotCleanup>{
```
Comment (mklechan) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java]: will make it private

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java (hunk 2 above)
Comment (mklechan) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java]: :+1:

File: h2o-py/h2o/frame.py (hunk 4 above)
Comment (mklechan) [inline h2o-py/h2o/frame.py]: also should be enum, time, or int, i'll add a comment

File: h2o-py/h2o/frame.py (hunk 5 above)
Comment (mklechan) [inline h2o-py/h2o/frame.py]: yes these are all column labels, havent handled numeric indexes...

File: h2o-py/tests/testdir_misc/pyunit_pivot_medium.py (hunk 6 above)
Comment (mklechan) [inline h2o-py/tests/testdir_misc/pyunit_pivot_medium.py]: for normal pivot usage, they can use whatever interval of time they want as long as its millis or larger. Here im chopping off the HH:MM:ss part because this is a random frame and without this step the result would be very sparse and not useful to analyze.

File: h2o-py/tests/testdir_misc/pyunit_pivot_medium.py (hunk 7 above)
Comment (mklechan) [inline h2o-py/tests/testdir_misc/pyunit_pivot_medium.py]: :+1:

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java (hunk 12 above)
Comment (mklechan) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java]: This was pretty tough to address, because the sort step does some chunk consolidation, so it was hard to make a test case where this loop had to go to the next chunk.  
Nevertheless, the new loop will update the next Chunk when necessary, and keeps going until the next index is different or the end of Vec has been reached.

File: h2o-r/tests/testdir_munging/runit_pivot.R (hunk 10 above)
Comment (mklechan) [inline h2o-r/tests/testdir_munging/runit_pivot.R]: I think it would be bad to reduplicate the same tests in both Python and R, because of redundancy and bogging down of the test system. I just have the R test here to test the client.  And the Python tests works on more of the bigger test sets and corner cases

Comment (navdeep-G) [inline h2o-r/tests/testdir_munging/runit_pivot.R]: I don't think it would bog down the test system. We need to ensure both clients behave the same. The test does not have to take a long time. Rather, it just has to ensure behavior between clients are the same.

File: h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java (hunk 12 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstPivot.java]: Thanks, I think this will work
